include/revng/PipeboxCommon/Concepts.h

```diff
@@ -89,35 +89,54 @@ concept IsAnalysis = requires(T &A) {
 namespace detail {
 
 template<typename T>
-struct PipeRunTraits {};
+concept HasContainerTypes = SpecializationOf<typename T::ContainerTypes,
+                                             TypeList>;
+
+template<typename T>
+struct PipeRunTraitsHelper {};
 
 template<typename C, typename... Args>
-  requires(IsContainerReference<Args> and ...)
-struct PipeRunTraits<
+  requires(not HasContainerTypes<C>) and (IsContainerReference<Args> and ...)
+struct PipeRunTraitsHelper<
   revng::pypeline::ObjectDependencies (C::*)(const Model &,
                                              const revng::pypeline::Request &,
                                              const revng::pypeline::Request &,
                                              llvm::StringRef,
                                              Args...)> {
-  using ContainerTypes = std::tuple<std::remove_reference_t<Args>...>;
+  using ContainerTypes = TypeList<std::remove_reference_t<Args>...>;
   static constexpr size_t Size = sizeof...(Args);
 };
 
+template<typename T>
```

Comment:
Describe the difference between a "pipe" (i.e., a long-lived object that should *not* have state) and a "pipe run" (i.e., an object whose alive while running the pipe and therefore you can store state that's specific to that run).

Reply:
Do you mean adding doxygen around? Where should I put it?